README
XML::RAI (RSS Abstraction Interface, said "ray") is an
object-oriented layer that maps overlapping and alternate tags in
RSS to one common simplified interface. RAI also includes an
liberal RSS parser that can be used seperately from the abstraction
layer.

INSTALLATION

To install this module type the following:

 perl Makefile.PL 
 make 
 make test 
 make install
 
DEPENDENCIES

This module requires these other modules and libraries:

XML::RSS::Parser 4.0
Date::Parse 2.26
Date::Format 2.22


COPYRIGHT AND LICENCE

The software is released under the Artistic License. The terms of
the Artistic License are described at
http://www.perl.com/language/misc/Artistic.html.

Except where otherwise noted, XML::RAI is Copyright
2003-2009, Timothy Appnel, tima@cpan.org. All rights
reserved.



Changes
Revision history for Perl extension XML::RAI.

1.303 Aug 13 2009
	- Applied memory leak patch from Mark Glines (mark@glines.org) Thanks!
	
1.3022July 18 2007
    - Changed XML::RAI::Object's date handler to avoid an apparent bug in 
      Date::Format::strftime. Thanks to Tim Sweetman.

1.301Nov 20 2005
    - Fixed Makefile dependencies

1.3  Nov 18 2005
    - Seperated out XML::RSS::Parser reversing a bad idea.
    - Introduced add_mapping method to all objects.
    - Now requires XML::RSS::Parser 4.0.
    - Various mapping tweaks.
    - Numerous bug fixes.
    - Test. A lot of them.

1.21 Jul 16 2005
    - Replaced use of Style param in XML::RSS::Parser constructor with
      standard Handlers param. This lowered the prerequisite version of 
      XML::Parser needed.

1.2  Jan 16 2005
    - POTENTIAL BREAK: XML::RAI::Object::source has been renamed src. This fixes
      a boneheaded oversight where XML::RAI::Item had two documented source
      methods. The source method for retreiving any RSS source attributes
      was overloading the XML::RAI::Object source method to retreive the 
      underlying XML::RSS::Parser::Element object. This was completely fouling 
      up retreival in an ARRAY context for all the RSS methods in the
      XML::RAI::Item class. This change required numerous modifications 
      throughout the XML::RAI classes to reflect this naming change.

1.11 Jan 09 2005
    - Fixed condition that was generating "Use of uninitialized value in 
      localtime" warning.

1.1  Jan 09 2005
    - Dropped POSIX dependency in favor of Date::Format. 
    - Added EPOCH time formatting option
    - Fixed time_format bug were PASS_THRU could not be set. (Kevin Frost)
    - Fixed PASS_THRU bug were an XML::RSS::Parser::Element object would be 
      returned instead of a string. (Kevin Frost)
    - Fixed bug were date normalization was ignoring timezone and using the 
      local system timezone instead. (Kevin Frost)

1.04 Nov 22 2004
    - Added the Technorati API namespace to XML::Parser::Element.

1.03 Nov 06 2004
    - Fixed RAI->item_count bug that was returning one less then the 
      actual count. (Jason Swan)

1.02 Oct 16 2004
    - Added missing Class::XPath dependency to Makefile.PL  

1.01 Sep 30 2004
    - Corrected XML::RSS::Parser prerequisite version to 3.
    - Added prerequisite version of Date::Parse to 2.26 in RAI::Object.
    - Switched use of deprecated value method to text_content in 
      RAI::Object.

1.00 Sep 29 2004
    - Integrated XML::RSS::Parser package into RAI.
    - Refactored XML::RSS::Parser to use XML::Elemental package and
      true tree structure.
    - Introduced XML::RSS::Parser::Characters class
    - Exposed internal methods for use with Class::XPath in Element
    - Added more graceful handling of unrecognized namespace URIs
    - Deprecated value and append_value. BREAKAGE: Regretfully value method 
      will not work the same as version 2x of the RSS Parser in some cases.
      This is due to the adoption of a tree model and how text is handled.
    - Added as_xml method for reimplementing the version 2x pass-thru 
      functionality under the tree structure.

0.52 Aug 14 2004
    - Removed lastBuildDate from the RAI::Item created and modified mappings. A
      good idea gone bad if you are trying to merge feeds and list out entries 
      chronological order.
    - Added minimum version of 0 to POSIX an Date::Parse in Makefile.PL.
    - Time handlers (created, modified etc) return UTC rather then local time

0.51 Jun 09 2004
    - Fixed multiple feed images bug that caused RAI to choke. (Mike Axelrod)
    - Add rdf:resource to image URL mapping.

0.50 May 03 2004
    - Added special link handlers to make up for Class::XPath's (1.4)
      lack of not() function and compound logical statements.
    - Fixed guid mapping. Added link module to mappings.
    - Added created_strict, issued_strict, and modified_strict methods
      to Element and Channel.
    - Class::XPath not liking 'description[@type="text/html"]'. 
      Removed from content_strict mapping for now.
    - Fixed deprecated contentstrict support.

0.41 Apr 28 2004
    - Fixed bad XPath statement in item/content_strict.

0.4  Apr 27  2004
    - Added content_strict. Deprecated contentstrict.
    - Modified mappings of channel/issued, channel/modified, relation,
      item/content, item/content_strict, item/creator, source.
    - Switched isPermalink to isPermaLink in link mapping.
    - Added pinged method to item. 
    - Folded channel/alt_table into description.
    - Fixed time_handler bug were no matches were not being caught 
      and handled.
    
0.3  Apr 18 2004
    - Added ability to instaniate RAI with an existing 
      XML::RSS::Parser tree.
    - Promoted @xml:lang in Channel and Item
    - Added Trackback "RSS 2.0" support in Item
    - Added identifier rdf:resource support in Item
    - Added dcterms:identifier to identifier in Item.
    - Added alt_title method to Channel
    
0.2  Apr 06 2004
    - Added time_format method to RAI and timestamp normalization 
      to Object.
    - Added a created method to Item
    - Added a ping method to Item
    - Added trackback:about to the related mapping of Item
    - Added xml:lang to the language mapping of Item
    - Missing documentation inserted
    - Dumped shared channel/item mapping. Each have their own
      separate maps now.

0.1  Mar 12 2004
	- Initial release into CPAN.


lib/XML/RAI/Item.pm
# Copyright (c) 2004-2009 Timothy Appnel
# http://appnel.com/
# This code is released under the Artistic License.
#
# XML::RAI::Item - an interface to the item elements in a RSS feed.
#

package XML::RAI::Item;

use strict;
use XML::RAI::Object;

use vars qw(@ISA $XMap);
@ISA = qw( XML::RAI::Object );

$XMap = {
    'format' => ['dc:format'],

#'link'=>['link','@rdf:about','guid[@isPermaLink="true"]'], # use special handler.
    abstract       => ['dcterms:abstract', 'description', 'dc:description'],
    content_strict => ['xhtml:body',       'xhtml:div',   'content:encoded'],
    content        => [
                'xhtml:body',      'xhtml:div',
                'content:encoded', 'description',
                'dc:description',  'rss091:description'
    ],
    contentstrict => ['xhtml:body', 'xhtml:div', 'content:encoded']
    ,    # deprecated
    contributor    => ['dc:contributor'],
    coverage       => ['dc:coverage'],
    created_strict => ['dcterms:created'],
    created => ['dcterms:created', 'dc:date', 'pubDate', 'rss091:pubDate'],
    creator     => ['dc:creator',  'author'],
    description => ['description', 'dc:description', 'dcterms:abstract'],
    identifier =>
      ['dc:identifier/@rdf:resource', 'dc:identifier', 'guid', 'link'],
    issued_strict => ['dcterms:issued'],
    issued        => ['dcterms:issued', 'dc:date', 'pubDate', 'rss091:pubDate'],
    language      => [
                 '@xml:lang',         'dc:language',
                 '/@xml:lang',        '/channel/dc:language',
                 '/channel/language', '/channel/rss091:language'
    ],
    modified_strict => ['dcterms:modified'],
    modified => ['dcterms:modified', 'dc:date', 'pubDate', 'rss091:pubDate'],
    ping      => ['trackback:ping/@rdf:resource',  'trackback:ping'],
    pinged    => ['trackback:about/@rdf:resource', 'trackback:about'],
    publisher => [
                  'dc:publisher',            '/channel/dc:publisher',
                  '/channel/managingEditor', '/channel/rss091:managingEditor'
    ],
    relation => ['dc:relation/@rdf:resource', 'dc:relation'],
    rights   => [
               'dc:rights',
               '/channel/copyright',
               '/channel/creativeCommons:license',
               '/channel/rss091:copyright'
    ],
    source  => ['dc:source',  'source/@url', 'source'],
    subject => ['dc:subject', 'category'],
    title   => ['title',      'dc:title'],
    type    => ['dc:type'],
    valid => ['dcterms:valid', 'expirationDate']
};

# Class::XPath is missing some functionality we need here so we
# help it along.
sub link {
    my $this = shift;
    my @nodes;

    # awkward use, but achieves the effect we need.
    if    (@nodes = $this->src->query('link'))       { }
    elsif (@nodes = $this->src->query('@rdf:about')) { }
    elsif (
        @nodes = grep {
            !$_->attributes
              || (!$_->attributes->{isPermaLink}
                  || $_->attributes->{isPermaLink} eq 'true')
        } $this->src->query('guid')
      ) {
      } elsif (
        @nodes = grep {
            $_->attributes->{type} =~ m!^(text/html|application/xhtml+xml)$!
        } $this->src->query('l:link[@rel="permalink"]')
      ) {
      } elsif (@nodes = $this->src->query('comment')) {
    }
    my @n = map { ref($_) ? $_->text_content : $_ } @nodes;
    wantarray ? @n : $n[0];
}

1;

__END__

=begin

=head1 NAME

XML::RAI::Item - An interface to the item elements of a RSS feed.

=head1 DESCRIPTION

A subclass of L<XML::RAI::Object>, XML::RAI::Item handles the
mapping function and retrieval of RSS item elements.

=head1 METHODS

=item $item->src

Returns the L<XML::RSS::Parser::Element> that the object is using
as its source. 

=item $item->parent

Returns the parent of the RAI object.

=item $item->add_mapping(key, @xpaths)

Creates or appends XPath mappings to the item object for
extensibility and easier access of RAI.



=head2 META DATA ACCESSORS

These accessor methods attempt to retrieve meta data from the
source L<XML::RSS::Parser> element by checking a list of potential
tag names until one returns a value.  They are generally based on
Dublin Core terminology and RSS elements that are common across the
many formats. If called in a SCALAR context, the value of the first
element of the tag being matched is returned. If called in an ARRAY
context it will return all of the values to the tag being matched
-- it does not return all of the values for all of the tags that
have been mapped to the method. (Note that some mappings only allow
one value to be returned.) Returns C<undef> if nothing could be
found.

The following are the tags (listed in XPath notation) mapped to
each method and the order in which they are checked.

=over 4

=item $item->abstract

=over 4

=item * dcterms:abstract

=item * description

=item * dc:description

=back

=item $item->content

=over 4

=item * xhtml:body

=item * xhtml:div

=item * content:encoded

=item * description

=item * dc:description

=item * rss091:description

=back

=item $item->content_strict

=over 4

=item * xhtml:body

=item * xhtml:div

=item * content:encoded

=back

=item $item->contentstrict

=over 4

=item * xhtml:body

=item * xhtml:div

=item * content:encoded

=back

=item $item->contributor

=over 4

=item * dc:contributor

=back

=item $item->coverage

=over 4

=item * dc:coverage

=back

=item $item->created

=over 4

=item * dcterms:created

=item * dc:date

=item * pubDate

=item * rss091:pubDate

=back

=item $item->created_strict

=over 4

=item * dcterms:created

=back

=item $item->creator

=over 4

=item * dc:creator

=item * author

=back

=item $item->description

=over 4

=item * description

=item * dc:description

=item * dcterms:abstract

=back

=item $item->format

=over 4

=item * dc:format

=back

=item $item->identifier

=over 4

=item * dc:identifier/@rdf:resource

=item * dc:identifier

=item * guid

=item * link

=back

=item $item->issued

=over 4

=item * dcterms:issued

=item * dc:date

=item * pubDate

=item * rss091:pubDate

=back

=item $item->issued_strict

=over 4

=item * dcterms:issued

=back

=item $item->language

=over 4

=item * @xml:lang

=item * dc:language

=item * /@xml:lang

=item * /channel/dc:language

=item * /channel/language

=item * /channel/rss091:language

=back

=item $item->link

=over 4

=item * link

=item * @rdf:about

=item * guid[@isPermaLink="true"]

=back

=item $item->modified

=over 4

=item * dcterms:modified

=item * dc:date

=item * pubDate

=item * rss091:pubDate

=back

=item $item->modified_strict

=over 4

=item * dcterms:modified

=back

=item $item->ping

=over 4

=item * trackback:ping/@rdf:resource

=item * trackback:ping

=back

=item $item->pinged

=over 4

=item * trackback:about/@rdf:resource

=item * trackback:about

=back

=item $item->publisher

=over 4

=item * dc:publisher

=item * /channel/dc:publisher

=item * /channel/managingEditor

=item * /channel/rss091:managingEditor

=back

=item $item->relation

=over 4

=item * dc:relation/@rdf:resource

=item * dc:relation

=back

=item $item->rights

=over 4

=item * dc:rights

=item * /channel/copyright

=item * /channel/creativeCommons:license

=item * /channel/rss091:copyright

=back

=item $item->source

=over 4

=item * dc:source

=item * source

=back

=item $item->subject

=over 4

=item * dc:subject

=item * category

=back

=item $item->title

=over 4

=item * title

=item * dc:title

=back

=item $item->type

=over 4

=item * dc:type

=back

=item $item->valid

=over 4

=item * dcterms:valid

=item * expirationDate

=back

The C<ping> and C<pinged> methods have been moved into their
own RAI plugin. See L<XML::RAI::TrackBack> for more.

=head1 AUTHOR & COPYRIGHT

Please see the XML::RAI manpage for author, copyright, and license
information.

=cut

=end


lib/XML/RAI/Image.pm
# Copyright (c) 2004-2009 Timothy Appnel
# http://appnel.com/
# This code is released under the Artistic License.
#
# XML::RAI::Image - An interface to the image elements of a RSS feed.
#

package XML::RAI::Image;

use strict;

use vars qw(@ISA $XMap);
@ISA = qw( XML::RAI::Object );

$XMap = {
    description =>
      ['/channel/image[0]/description', '/channel/image[0]/rss091:description'],
    height => ['/channel/image[0]/height', '/channel/image[0]/rss091:height'],
    'link' => ['/channel/image[0]/link'],
    title  => ['/channel/image[0]/title',  '/channel/image[0]/dc:title'],
    url    => [
            '/channel/image[0]/@rdf:resource', '/channel/image[0]/@rdf:about',
            '/channel/image[0]/url'
    ],
    width => ['/channel/image[0]/width', '/channel/image[0]/rss091:width']
};

1;

__END__

=begin

=head1 NAME

XML::RAI::Image - An interface to the image elements of a RSS feed.

=head1 DESCRIPTION

A subclass of L<XML::RAI::Object>, XML::RAI::Image handles the
mapping function and retrieval of RSS channel elements.

=head1 METHODS

=item $image->src

Returns the L<XML::RSS::Parser::Element> that the object is using
as its source.

=item $image->parent

Returns the parent of the RAI object.

=item $image->add_mapping(key, @xpaths)

Creates or appends XPath mappings to the image object for
extensibility and easier access of RAI.

=head2 META DATA ACCESSORS

These accessor methods attempt to retrieve meta data from the
source L<XML::RSS::Parser> element by checking a list of potential
tag names until one returns a value.  They are generally based on
Dublin Core terminology and RSS elements that are common across the
many formats. If called in a SCALAR context, the value of the first
element of the tag being matched is returned. If called in an ARRAY
context it will return all of the values to the tag being matched
-- it does not return all of the values for all of the tags that
have been mapped to the method. (Note that some mappings only allow
one value to be returned.) Returns C<undef> if nothing could be
found.

The following are the tags (listed in XPath notation) mapped to
each method and the order in which they are checked.

=over 4

=item $image->description

=over 4

=item * /channel/image[0]/description

=item * /channel/image[0]/rss091:description

=back

=item $image->height

=over 4

=item * /channel/image[0]/height

=item * /channel/image[0]/rss091:height

=back

=item $image->link

=over 4

=item * /channel/image[0]/link

=back

=item $image->title

=over 4

=item * /channel/image[0]/title

=item * /channel/image[0]/dc:title

=back

=item $image->url

=over 4

=item * /channel/image[0]/@rdf:resource

=item * /channel/image[0]/@rdf:about

=item * /channel/image[0]/url

=back

=item $image->width

=over 4

=item * /channel/image[0]/width

=item * /channel/image[0]/rss091:width

=back

=head1 AUTHOR & COPYRIGHT

Please see the XML::RAI manpage for author, copyright, and license
information.

=cut

=end


lib/XML/RAI/Object.pm
# Copyright (c) 2004-2009 Timothy Appnel
# http://appnel.com/
# This code is released under the Artistic License.
#
# XML::RAI::Object - A base class for RAI element objects.
#

package XML::RAI::Object;

use strict;

use Date::Parse 2.26;
use Date::Format 2.22;
use Scalar::Util qw(weaken);

sub new {
    my $class = shift;
    my $self = bless {}, $class;
    $self->init(@_);
    $self;
}

sub init {
    $_[0]->{__source} = $_[1];
    $_[0]->{__parent} = $_[2];
    $_[0]->{__RAI}    = $_[2];
    while ($_[0]->{__RAI}->can('parent')) {
        $_[0]->{__RAI} = $_[0]->{__RAI}->parent;
    }
    weaken($_[0]->{__parent});
    weaken($_[0]->{__RAI});
}

sub src    { $_[0]->{__source} }
sub parent { $_[0]->{__parent} }

sub add_mapping {
    my ($class, $var, @maps) = @_;
    no strict 'refs';
    ${"${class}::XMap"}->{$var} = []
      unless (exists ${"${class}::XMap"}->{$var});
    push @{${"${class}::XMap"}->{$var}}, @maps;
}

sub generic_handler {
    my ($this, $class, $var) = @_;
    no strict 'refs';
    foreach (@{${$class . '::XMap'}->{$var}}) {
        my @nodes = $this->src->query($_);
        if (defined($nodes[0])) {
            @nodes = map {
                !ref($_) ? $_
                  : substr($_->name, 0, 30) eq '{http://www.w3.org/1999/xhtml}'
                  ? join '', map { as_xhtml($_) } @{$_->contents}
                  : $_->text_content;
            } @nodes;
            return wantarray ? @nodes : $nodes[0];
        }
    }
    return undef;
}

sub time_handler {
    my @r = generic_handler(@_);
    return undef unless $r[0];
    my $timef = $_[0]->{__RAI}->time_format;
    if ($timef eq 'EPOCH') {
        map { $_ = str2time($_, 0) } @r;
    } elsif ($timef) {
        map { $_ = time2str($timef, str2time($_, 0), 0) } @r;
    }
    wantarray ? @r : $r[0];
}

use XML::RSS::Parser::Util qw(as_xml);

sub as_xhtml {    # a hack to get xhtml content as we'd expect.
    my $xml = as_xml($_[0]);
    $xml =~ s{<(/?)xhtml:}{<$1}g;
    $xml;
}

sub DESTROY { }

use vars qw( $AUTOLOAD );

sub AUTOLOAD {
    (my $var   = $AUTOLOAD) =~ s!.+::!!;
    (my $class = $AUTOLOAD) =~ s!::[^:]+$!!;
    no strict 'refs';
    die "$var is not a recognized method."
      unless (${$class . '::XMap'}->{$var});
    if ($var =~ m/^(created|modified|issued|valid)(_strict)?$/) {
        *$AUTOLOAD = sub { time_handler($_[0], $class, $var) };
    } else {
        *$AUTOLOAD = sub { generic_handler($_[0], $class, $var) };
    }
    goto &$AUTOLOAD;
}

1;

__END__

=begin

=head1 NAME

XML::RAI::Object - A base class for RAI element objects.

=head1 DESCRIPTION

XML::RAI::Object is an base class for RAI element objects.
Subclasses need only plug in an element map (C<$I<PACKAGE>::XMap>)
consisting of method names (the key) and a list of XPath-esque
queries to use as a search path in the RSS parse tree.

=head1 AUTHOR & COPYRIGHT

Please see the XML::RAI manpage for author, copyright, and license
information.

=cut

=end


lib/XML/RAI/Channel.pm
# Copyright (c) 2004-2009 Timothy Appnel
# http://appnel.com/
# This code is released under the Artistic License.
#
# XML::RAI::Channel - an interface to the channel elements of a RSS feed.
#

package XML::RAI::Channel;

use strict;
use XML::RAI::Object;

use vars qw(@ISA $XMap);
@ISA = qw( XML::RAI::Object );

use XML::RAI;

$XMap = {
    'format' => ['dc:format'],

    # 'link'=>['link','@rdf:about'], #special handler
    contributor => ['dc:contributor'],
    coverage    => ['dc:coverage'],
    creator     => ['dc:creator'],
    description => [
                    'description',      'dc:description',
                    'dcterms:abstract', 'dcterms:alternative'
    ],
    generator => [
                  'admin:generatorAgent/@rdf:resource',
                  'admin:generatorAgent',
                  'generator'
    ],
    identifier    => ['dc:identifier/@rdf:resource', 'dc:identifier', 'link'],
    issued_strict => ['dcterms:issued'],
    issued        =>
      ['dcterms:issued', 'dc:date', 'lastBuildDate', 'rss091:lastBuildDate'],
    language   => ['@xml:lang', 'dc:language', 'language', 'rss091:language'],
    maintainer => [
                   'admin:errorReportsTo/@rdf:resource',
                   'admin:errorReportsTo',
                   'webMaster'
    ],
    modified_strict => ['dcterms:modified'],
    modified        =>
      ['dcterms:modified', 'dc:date', 'lastBuildDate', 'rss091:lastBuildDate',],
    publisher => ['dc:publisher', 'managingEditor', 'rss091:managingEditor'],
    relation => ['dc:relation/@rdf:resource', 'dc:relation'],
    rights   =>
      ['dc:rights', 'copyright', 'creativeCommons:license', 'rss091:copyright'],
    source  => ['dc:source',  'source/@url', 'source', 'title'],
    subject => ['dc:subject', 'category'],
    title   => ['title',      'dc:title'],
    type    => ['dc:type'],
    valid => ['dcterms:valid', 'expirationDate'],
};

# Class::XPath is missing some functionality we need here so we
# help it along.
sub link {
    my $this = shift;
    my @nodes;

    # awkward use, but achieves the effect we need.
    if (@nodes = $this->src->query('link')) { }
    elsif (
        @nodes = grep {
            $_->attributes->{type} =~ m!^(text/html|application/xhtml+xml)$!
        } $this->src->query('l:link[@rel="permalink"]')
      ) {
      } elsif (@nodes = $this->src->query('dc:relation/@rdf:resource')) {
      } elsif (@nodes = $this->src->query('dc:relation')) {
    }
    return unless (defined $nodes[0]);
    my @n = map { ref($_) ? $_->text_content : $_ } @nodes;
    wantarray ? @n : $n[0];
}

1;

__END__

=begin

=head1 NAME

XML::RAI::Channel - An interface to the channel elements of a RSS
feed.

=head1 DESCRIPTION

A subclass of L<XML::RAI::Object>, XML::RAI::Channel handles the
mapping function and retrieval of RSS channel elements.

=head1 METHODS

=item $channel->src

Returns the L<XML::RSS::Parser::Element> that the object is using
as its source.

=item $channel->parent

Returns the parent of the RAI object.

=item $channel->add_mapping(key, @xpaths)

Creates or appends XPath mappings to the channel object for extensibility
and easier access of RAI.

=head2 META DATA ACCESSORS

These accessor methods attempt to retrieve meta data from the
source L<XML::RSS::Parser> element by checking a list of potential
tag names until one returns a value.  They are generally based on
Dublin Core terminology and RSS elements that are common across the
many formats. If called in a SCALAR context, the value of the first
element of the tag being matched is returned. If called in an ARRAY
context it will return all of the values to the tag being matched
-- it does not return all of the values for all of the tags that
have been mapped to the method. (Note that some mappings only allow
one value to be returned.) Returns C<undef> if nothing could be
found.

The following are the tags (listed in XPath notation) mapped to
each method and the order in which they are checked.

=over 4

=item $channel->contributor

=over 4

=item * dc:contributor

=back

=item $channel->coverage

=over 4

=item * dc:coverage

=back

=item $channel->creator

=over 4

=item * dc:creator

=back

=item $channel->description

=over 4

=item * description

=item * dc:description

=item * dcterms:abstract

=item * dcterms:alternative

=back

=item $channel->format

=over 4

=item * dc:format

=back

=item $channel->generator

=over 4

=item * admin:generatorAgent/@rdf:resource

=item * admin:generatorAgent

=item * generator

=back

=item $channel->identifier

=over 4

=item * dc:identifier/@rdf:resource

=item * dc:identifier

=item * link

=back

=item $channel->issued

=over 4

=item * dcterms:issued

=item * dc:date

=item * lastBuildDate

=item * rss091:lastBuildDate

=back

=item $channel->issued_strict

=over 4

=item * dcterms:issued

=back

=item $channel->language

=over 4

=item * @xml:lang

=item * dc:language

=item * language

=item * rss091:language

=item $channel->link

=over 4

=item * link

=item * @rdf:about

=back

=item $channel->maintainer

=over 4

=item * admin:errorReportsTo/@rdf:resource

=item * admin:errorReportsTo

=item * webMaster

=back

=item $channel->modified

=over 4

=item * dcterms:modified

=item * dc:date

=item * lastBuildDate

=item * rss091:lastBuildDate

=back

=item $channel->modified_strict

=over 4

=item * dcterms:modified

=back

=item $channel->publisher

=over 4

=item * dc:publisher

=item * managingEditor

=item * rss091:managingEditor

=back

=item $channel->relation

=over 4

=item * dc:relation/@rdf:resource

=item * dc:relation

=back

=item $channel->rights

=over 4

=item * dc:rights

=item * copyright

=item * creativeCommons:license

=item * rss091:copyright

=back

=item $channel->source

=over 4

=item * dc:source

=item * source

=item * title

=back

=item $channel->subject

=over 4

=item * dc:subject

=item * category

=back

=item $channel->title

=over 4

=item * title

=item * dc:title

=back

=item $channel->type

=over 4

=item * dc:type

=back

=item $channel->valid

=over 4

=item * dcterms:valid

=item * expirationDate

=back

=back

=head1 AUTHOR & COPYRIGHT

Please see the XML::RAI manpage for author, copyright, and license
information.

=cut

=end


lib/XML/RAI.pm
# Copyright (c) 2004-2009 Timothy Appnel
# http://appnel.com/
# This code is released under the Artistic License.
#
# XML::RAI - RSS Abstraction Interface.
#

package XML::RAI;

use strict;

use vars qw($VERSION);
$VERSION = 1.303;

use XML::RSS::Parser 4.0;
use XML::RAI::Channel;
use XML::RAI::Item;
use XML::RAI::Image;

use constant W3CDTF    => '%Y-%m-%dT%H:%M:%S%z';    # AKA...
use constant RFC8601   => W3CDTF;
use constant RFC822    => '%a, %d %b %G %T %Z';
use constant PASS_THRU => '';
use constant EPOCH     => 'EPOCH';

sub new {
    my $class = shift;
    my $self = bless {}, $class;
    $self->init(@_);
    $self;
}

sub init {
    my $self = shift;
    my $doc;
    unless (ref($_[0]) eq 'XML::RSS::Parser::Feed') {
        my ($method, @r) = @_;
        my $parser = XML::RSS::Parser->new;
        $doc = $parser->$method(@r) or die $parser->errstr;
    }
    else {
        $doc = shift;
    }
    $self->{__doc} = $doc;
    my $channel = $self->{__channel} =
      XML::RAI::Channel->new($doc->channel, $self);
    my @items = map { XML::RAI::Item->new($_, $channel) } $doc->items;
    $self->{__items} = \@items;
    my @imgs = $doc->image;    # fix multiple image bug ala slashdot.
    $self->{__image} = XML::RAI::Image->new($imgs[0], $channel)
      if $doc->image;
    $self->{__timef} = W3CDTF;
}

sub time_format {
    $_[0]->{__timef} = $_[1] if defined $_[1];
    $_[0]->{__timef};
}

sub parse {
    my $class = shift;
    if (ref($_[0]) eq 'GLOB') {    # is filehandle
        $class->parse_file(@_);
    }
    else {                         # is string
        $class->parse_string(@_);
    }
}

sub parsefile {
    my $class = shift;
    $class->new('parse_file', @_) or die $class->errstr;
}
*parse_file = \&parsefile;

sub parse_string {
    my $class = shift;
    $class->new('parse_string', @_) or die $class->errstr;
}

sub parse_uri {
    my $class = shift;
    $class->new('parse_uri', @_) or die $class->errstr;
}

sub document   { $_[0]->{__doc}; }
sub channel    { $_[0]->{__channel}; }
sub items      { $_[0]->{__items}; }
sub item_count { scalar @{$_[0]->{__items}}; }
sub image      { $_[0]->{__image}; }

1;

__END__

=begin

=head1 NAME

XML::RAI - RSS Abstraction Interface.

=head1 SYNOPSIS

 #!/usr/bin/perl -w
 use strict;
 use XML::RAI;
 my $doc = <<DOC;
 <?xml version="1.0" encoding="iso-8859-1"?>
 <rss xmlns:dc="http://purl.org/dc/elements/1.1/"
     xmlns="http://purl.org/rss/1.0/">
     <channel>
         <title>tima thinking outloud</title>
         <link>http://www.timaoutloud.org/</link>
         <description></description>
         <dc:language>en-us</dc:language>
         <item>
             <title>His and Hers Weblogs.</title>
             <description>First it was his and hers Powerbooks. Now 
             its weblogs. There goes the neighborhood.</description>
             <link>http://www.timaoutloud.org/archives/000338.html</link>
             <dc:subject>Musings</dc:subject>
             <dc:creator>tima</dc:creator>
             <dc:date>2004-01-23T12:33:22-05:00</dc:date>
         </item>
         <item>
             <title>Commercial Music Again.</title>
             <description>Last year I made a post about music used 
             in TV commercials that I recognized and have been listening to. 
             For all the posts I made about technology and other bits of sagely
             wisdom the one on commercial music got the most traffic of any 
             each month. I need a new top post. Here are some more tunes that 
             have appeared in commercials.</description>
             <guid isPermaLink="true">
               http://www.timaoutloud.org/archives/000337.html
             </guid>
             <category>Musings</category>
             <author>tima</author>
             <pubDate>Sun, 18 Jan 2004 14:09:03 GMT</pubDate>
         </item>
     </channel>
 </rss>
 DOC

 # The above is to demonstrate the value of RAI. It is not any 
 # specific RSS format, nor does it exercise best practices.

 my $rai = XML::RAI->parse_string($doc);
 print $rai->channel->title."\n\n";
 foreach my $item ( @{$rai->items} ) {
    print $item->title."\n";
    print $item->link."\n";
    print $item->content."\n";
    print $item->issued."\n\n";
 }

=head1 DESCRIPTION

The RSS Abstraction Interface, or RAI (said "ray"), provides an
object-oriented interface to XML::RSS::Parser trees that abstracts
the user from handling namespaces, overlapping and alternate tag
mappings.

It's rather well known that, while popular, the RSS syntax is a bit
of a mess. Anyone who has attempted to write software that consumes
RSS feeds "in the wild" can attest to the headaches in handling the
many formats and interpretations that are in use. For instance, in
"The myth of RSS compatibility"
L<http://diveintomark.org/archives/2004/02/04/incompatible-rss>
Mark Pilgrim identifies 9 different versions of RSS (there are 10
actually[1]) and that is not without going into tags with
overlapping purposes. Even the acronym RSS has multiple though
similar meanings.

The L<XML::RSS::Parser> alone attempts to help developers cope with these
issues through a liberal interpretation of what is RSS and routines
to normalize the parse tree into a more common and manageable form.

RAI takes this one step further. Its intent is to give a developer
the means to not have to care about what tags the feed uses to
present its meta data.

RAI provides a single simplified interface that maps one method
call to various overlapping and alternate tags used in RSS feeds.
The interface also abstracts developers from needing to deal with
namespaces. Method names are based on Dublin Core terminology.

With the release of version 1.0, the L<XML::RSS::Parser> 
distribution was folded into XML::RAI. 

[1] When initially released, RSS 2.0 had a namespace. When it was
reported a few days later that some XSLT-based systems were
breaking because of the change in the RSS namespace from "" (none)
to http://backend.userland.com/rss2, the namespace was removed, but
the version number was not incremented making it incompatible with
itself. L<http://groups.yahoo.com/group/rss-dev/message/4113> This
version was not counted in Mark's post.

=head1 METHODS

=item XML::RAI->new($rss_tree)

Returns a populated RAI instance based on the 
L<XML::RSS::Parser::Feed> object passed in.

=item XML::RAI->parse($string_or_file_handle)

Passes through the string or file handle to the C<parse>
method to either C<parse_file> or C<parse_string> in
L<XML::RSS::Parser>. Returns a populated RAI instance.

To maintain backwards compatability this method is B<not> inherited 
from the underlying SAX implementation.

=item XML::RAI->parse_file

=item XML::RAI->parse_string

=item XML::RAI->parse_uri

A pass-thru to the underlying SAX implentation. See L<XML::SAX::Base> for 
more on these methods.

=item $rai->document

Returns the L<XML::RSS::Parser> parse tree being used as the source
for the RAI object

=item $rai->channel

Returns the L<XML::RAI::Channel> object.

=item $rai->items

Returns an array reference containing the L<XML::RAI::Item> objects
for the feed

=item $rai->item_count

Returns the number of items as an integer.

=item $rai->image

Returns the L<XML::RAI::Image> object, if any. (Many feeds do not
have an image block.)

=item $rai->time_format($timef)

Sets the timestamp normalization format. RAI will attempt to parse
the string into a data value and will output timestamp (date)
values in this format.

RAI implements a few constants with common RSS timestamp formatting
strings:

 W3CDTF     1999-09-01T22:10:40Z 
 RFC8601    (other name for W3CDTF)
 RFC822     Wed, 01 Sep 1999 22:10:40 GMT 
 EPOCH      (Seconds since system epoch.)
 PASS_THRU  (timestamp as it appear in the source. does not normalize.)

W3CDTF/RFC8601 is the default. For more detail on creating your own
timestamp formats see the manpage for the C<strftime> command.

=head1 PLUGINS

With the introduction of the C<add_mapping> and the
C<register_ns_prefix> method in the underlying
L<XML::RSS::Parser>, RAI now has a plugin API for easily
extending its mappings.

To create a RAI plugin module, simply create a package with
an C<import> method that makes all of the necessary
C<add_mapping> and C<register_ns_prefix> calls. For an
example plugin module see L<XML::RAI::TrackBack>

=head1 DEPENDENCIES

L<XML::RSS::Parser> 4.0, L<Date::Parse> 2.26, L<Date::Format> 2.22

=head1 TO DO

=over

=item * Add Atom elements into mappings.

=item * Serialization module(s).

=item * DATETIME (L<DateTime> object) constants and functionality 
for C<time_format>.

=back

=head1 PARTICIPATION

I welcome and accept patches in diff format. If you wish to
hack on this code, please fork the git repository found at:
L<http://github.com/tima/perl-xml-rai/>

If you have something to push back to my repository, just
use the "pull request" button on the github site.

=head1 LICENSE

The software is released under the Artistic License. The terms of
the Artistic License are described at
L<http://www.perl.com/language/misc/Artistic.html>.

=head1 AUTHOR & COPYRIGHT

Except where otherwise noted, XML::RAI is Copyright
2003-2009, Timothy Appnel, tima@cpan.org. All rights
reserved.

=cut

=end
